test/org/sagebionetworks/bridge/models/schedules/IntervalActivitySchedulerTest.java

```diff
@@ -608,10 +610,91 @@ public void eventIsLateUTC() {
+        
+        String timestamp = firstTimeStampFor(-8, -8, schedule);
+        assertEquals("2017-02-19T23:00:00.000-08:00", timestamp);
+        
+        timestamp = firstTimeStampFor(9, 9, schedule);
+        assertEquals("2017-02-20T23:00:00.000+09:00", timestamp);
+    }
+    
+    // Scheduler is interpreting the event to the right local time. Examples:
+    // Enrollment=1487552400000 (2017-02-20T01:00Z), timeZone=-08:00, endsOnTimeZone=-08:00, schedule=one hour after enrollment, expected=2017-02-19T18:00-0800
+    // Enrollment=1487552400000 (2017-02-20T01:00Z), timeZone=+09:00, endsOnTimeZone=+09:00, schedule=one hour after enrollment, expected=2017-02-20T11:00+0900
+    @Test
+    public void schedulerInterpetsEventToRightLocalTime() {
+        Schedule schedule = new Schedule();
+        schedule.getActivities().add(TestConstants.TEST_1_ACTIVITY);
+        schedule.setScheduleType(ScheduleType.ONCE);
+        schedule.setDelay(Period.parse("PT1H"));
+        schedule.setExpires("P1D");
+        
+        String timestamp = firstTimeStampFor(-8, -8, schedule);
+        assertEquals("2017-02-19T18:00:00.000-08:00", timestamp);
+        
+        timestamp = firstTimeStampFor(9, 9, schedule);
+        assertEquals("2017-02-20T11:00:00.000+09:00", timestamp);
+    }
+
+    // Local times having timezone correctly applied. Examples:
+    // Enrollment=1487552400000 (2017-02-20T01:00Z), timeZone=-08:00, endsOnTimeZone=+09:00, schedule=one hour after enrollment, expected=2017-02-19T18:00+0900 (Should this be the 19th or the 20th?)
+    // Enrollment=1487552400000 (2017-02-20T01:00Z), timeZone=+09:00, endsOnTimeZone=-08:00, schedule=one hour after enrollment, expected=2017-02-20T11:00-0800
+    @Test
+    public void localTimeZonesCorrectlyApplied() {
+        Schedule schedule = new Schedule();
+        schedule.getActivities().add(TestConstants.TEST_1_ACTIVITY);
+        schedule.setScheduleType(ScheduleType.ONCE);
+        schedule.setDelay(Period.parse("PT1H"));
+        schedule.setExpires("P1D");
+        
+        String timestamp = firstTimeStampFor(-8, +9, schedule);
+        assertEquals("2017-02-19T18:00:00.000+09:00", timestamp);
+        
+        timestamp = firstTimeStampFor(9, -8, schedule);
+        assertEquals("2017-02-20T11:00:00.000-08:00", timestamp);
+    }
+    
+    private String firstTimeStampFor(int initialTZOffset, int requestTZOffset, Schedule schedule) {
+        DateTime enrollment = new DateTime(1487552400000L); // 2017-02-20T01:00Z
+        DateTimeZone initialTimeZone = DateTimeZone.forOffsetHours(initialTZOffset);
+        DateTimeZone requestTimeZone = DateTimeZone.forOffsetHours(requestTZOffset);
+        
+        // NOTE: The events map is adjusted to initial time zone in the service, we must do that
+        // here for this test to pass. Just calling scheduler without a "localized" event map will fail.
+        events.put("enrollment", enrollment.withZone(initialTimeZone));
```

Comment:
Is this because the logic for parsing the enrollment time lives in ScheduledActivityService, but this test is in ActivityScheduler? If so, I would argue that this test doesn't fully exercise our code. Rather, we make an assumption about how callers will behave, which may not be true if there's a bug in the caller.

We should move this test up to ScheduledActivityService to verify our fix works end-to-end.